### PR TITLE
Enhanced Astro codespace support

### PR DESCRIPTION
**Pull request by bdmac**
## Changes
See https://github.com/withastro/roadmap/discussions/472 for further context.

Enhances the base Codespaces devcontainer and Dockerfile to provide a smoother development experience out of the box for contributors. Users of the base devcontainer will now be dropped into a fully functional development environment in their codespace and tests will automatically start running for them in a terminal tab.

Adds a devcontainer per example so that we can add buttons on astro.new that will launch users straight into the appropriate example app. Additional minimal Dockerfiles were created to support general example apps and the deno example app. A devcontainer was created for each example to ensure that the user is launched into the appropriate folder, has the appropriate initial file open, and that the example app is running on their behalf once the codespace loads. The built in browser preview should also show the running example by default.

Screenshot 2023-02-17 at 2 50 00 PM

Screenshot 2023-02-17 at 2 26 22 PM

## Testing

Only tweaking Codespaces devcontainer setups so no tests were added or updated.

## Docs

I've adjusted the readmes in all examples to include links to starting a codespace on that example as shown above in the screenshots. After this is merged I can also update astro.new to add these same links.

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM mcr.microsoft.com/devcontainers/javascript-node
```
Comment (samruddhikhandale) [inline .devcontainer/Dockerfile:1]: Can we pin this image to a specific node version and probably reference a semantic version ? 
This way we could avoid inconsistencies, and build failures due to future (even breaking) updates to this image.

```suggestion
FROM mcr.microsoft.com/devcontainers/javascript-node:0-18
```

File: .devcontainer/basics/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      "codespaces": {
+        "openFiles": [
+          "src/pages/index.astro"
+        ]
+      },
+      "vscode": {
+        "extensions": [
+          "astro-build.astro-vscode",
+          "esbenp.prettier-vscode",
+          "dbaeumer.vscode-eslint"
```
Comment (samruddhikhandale) [inline .devcontainer/basics/devcontainer.json:38]: We could remove this as this extension is already added by the image
https://github.com/devcontainers/images/blob/main/src/javascript-node/.devcontainer/devcontainer.json#L27

File: .devcontainer/blog/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      "codespaces": {
+        "openFiles": [
+          "src/pages/index.astro"
+        ]
+      },
+      "vscode": {
+        "extensions": [
+          "astro-build.astro-vscode",
+          "esbenp.prettier-vscode",
+          "dbaeumer.vscode-eslint"
```
Comment (samruddhikhandale) [inline .devcontainer/blog/devcontainer.json:38]: Same comment here and for every other devcontainer.json which is based upon the javascript-node image

File: .devcontainer/examples.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,6 @@
+FROM mcr.microsoft.com/devcontainers/javascript-node
```
Comment (samruddhikhandale) [inline .devcontainer/examples.Dockerfile:1]: > Can we pin this image to a specific node version and probably reference a semantic version ?

Same comment here and for every other dockerfile

File: .devcontainer/basics/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
+// https://github.com/microsoft/vscode-dev-containers/tree/v0.238.1/containers/docker-existing-dockerfile
```
Comment (samruddhikhandale) [inline .devcontainer/basics/devcontainer.json:2]: > https://github.com/microsoft/vscode-dev-containers/tree/v0.238.1/containers/docker-existing-dockerfile

Wondering why would we need a reference to `docker-existing-dockerfile`? 🤔 
Also, the container in `microsoft/vscode-dev-containers` are not currently supported and are moved to the [dev containers](https://github.com/devcontainers) org.

File: .devcontainer/example-welcome-message.txt [hunk 6]
```diff
@@ -0,0 +1,3 @@
+👋 Welcome to Astro!
+
+The example app should automatically start soon in a new terminal tab.
```
Comment (samruddhikhandale) [inline .devcontainer/example-welcome-message.txt:3]:
```suggestion
👋 Welcome to "Astro" in GitHub Codespaces!

🛠️  Your environment is fully setup with all the required software.

🚀 The example app should automatically start soon in a new terminal tab.
```

Comment (samruddhikhandale) [reply .devcontainer/example-welcome-message.txt:3]: Probably we could make similar changes to other welcome messages as well?

**Review by samruddhikhandale — COMMENTED** [review]
This is awesome, thank you so much! ✨ 
Left some comments/thoughts.

File: .devcontainer/basics/devcontainer.json (hunk 5 above)
Comment (bdmac) [reply .devcontainer/basics/devcontainer.json:2]: Probably copied from somewhere.

File: .devcontainer/basics/devcontainer.json (hunk 2 above)
Comment (bdmac) [reply .devcontainer/basics/devcontainer.json:38]: Sure, I didn't even realize the images could _have_ their own devcontainers nor how they interact with ours!

File: .devcontainer/example-welcome-message.txt (hunk 6 above)
Comment (bdmac) [reply .devcontainer/example-welcome-message.txt:3]: Not following what you mean here.

Comment (samruddhikhandale) [reply .devcontainer/example-welcome-message.txt:3]: I am suggesting an update to the [example-welcome-message.txt](https://github.com/withastro/astro/pull/6288/files/7c4fce28e22b47fa85cca54600ac01d7a0737dc3#diff-c26282470ea314cfa922b5047b78b10fe214470a8284be71122f17c0ecab2c88) contents. If you decide to apply those changes, then you could also update other welcome files as well. (eg. devcontainer/welcome-message.txt)

**Review by natemoo-re — APPROVED** [review]
🙌 This is amazing, thanks so much for the PR!

This LGTM but I'd love one more review before merging.

**Review by Yan-Thomas — APPROVED** [review]
Docs LGTM, thanks @bdmac!

**Comment by natemoo-re** [comment]
Cool, I'm going to go ahead and merge this @bdmac! Thanks so much for the PR. 🙌

Regarding `astro.new`, we just finalized the mockups for the planned redesign, so we'll probably be overhauling that site early in Q2. You're still welcome to create a PR to add these, but I just wanted to give you a heads up that the look and feel will be changing significantly in a month or two.

**Pull request merged**